src/durmod/kaldi-durmod.h

```diff
@@ -0,0 +1,266 @@
+#include "util/stl-utils.h"
+#include "fstext/deterministic-fst.h"
+
+namespace kaldi {
+  using nnet3::Nnet;
+  using nnet3::NnetExample;
+  using nnet3::NnetIo;
+  using nnet3::CachingOptimizingCompiler;
+  using nnet3::ComputationRequest;
+  using nnet3::NnetComputation;
+  using nnet3::NnetComputeOptions;
+  using nnet3::NnetComputer;
+
+struct PhoneDurationModelOptions {
+  int32 left_context, right_context;
+  int32 max_duration;
+
+  void Register(OptionsItf *po);
+
+  PhoneDurationModelOptions():
+    left_context(4),
+    right_context(2),
+    max_duration(60) {}
+};
+
+class PhoneDurationModel {
+  friend class PhoneDurationFeatureMaker;
+
+ public:
+  void Read(std::istream &is, bool binary);
+  void Write(std::ostream &os, bool binary) const;
+
+  PhoneDurationModel() {}
+  PhoneDurationModel(const PhoneDurationModelOptions &opts,
+                     const std::vector<std::vector<int32> > &roots,
+                     const std::vector<std::vector<int32> > &questions):
+    left_context_(opts.left_context),
+    right_context_(opts.right_context),
+    roots_(roots),
+    questions_(questions),
+    max_duration_(opts.max_duration) {}
+
+  inline int32 MaxDuration() const { return max_duration_; }
+  inline int32 RightContext() const { return right_context_; }
+  inline int32 LeftContext() const { return left_context_; }
+  inline int32 FullContextSize() const {
+    return left_context_ + right_context_ + 1;
+  }
+  std::string Info() const;
+
+ private:
+  int32 left_context_, right_context_;
+  std::vector<std::vector<int32> > roots_;
+  std::vector<std::vector<int32> > questions_;
+  int32 max_duration_;
+};
+
+class PhoneDurationFeatureMaker {
+ public:
+  explicit PhoneDurationFeatureMaker(const PhoneDurationModel &model);
+
+  void InitFeatureMaker(const PhoneDurationModel &model);
+
+  /// This method extracts features for a phone. The inputs are a
+  /// phone-duration context and an index into that context which indicates the
+  /// middle phone (middle in the sense of left/right context).
+  /// a phone-duration context is a sequence of <phone-ID, duration-in-frames>
+  /// pairs.
+  /// The extracted features consist of phone-IDs (which are represented in a
+  /// 1-of-n encoding, suitable for neural networks), phone durations, and
+  /// binary features determined from a set of questions
+  /// (i.e. extra_questions.int)
+  void MakeFeatureVector(
+                 const std::vector<std::pair<int32, int32> > &phone_dur_context,
+                 int phone_index,
+                 SparseVector<BaseFloat> *feat) const;
+
+
+  int32 FeatureDim() const { return feature_dim_; }
+  int32 NumBinaryFeatures() const { return num_binary_features_; }
+  int32 NumPhoneIdentities() const { return num_phone_identities_; }
+  int32 OutputDim() const { return max_duration_; }
+
+  std::string Info() const;
+
+ private:
+  unordered_map<int32, std::vector<int32> > binary_feats_;
+  unordered_map<int32, int32> phone_id_;
```

Comment:
On Fri, Feb 19, 2016 at 5:52 PM, Hossein Hadian <notifications@github.com>
wrote:

> In src/durmod/kaldi-durmod.h
> <https://github.com/kaldi-asr/kaldi/pull/414#discussion_r53530457>:
>
> > +  void MakeFeatureVector(
> > +                 const std::vector<std::pair<int32, int32> > &phone_dur_context,
> > +                 int phone_index,
> > +                 SparseVector<BaseFloat> *feat) const;
> > +
> > +
> > +  int32 FeatureDim() const { return feature_dim_; }
> > +  int32 NumBinaryFeatures() const { return num_binary_features_; }
> > +  int32 NumPhoneIdentities() const { return num_phone_identities_; }
> > +  int32 OutputDim() const { return max_duration_; }
> > +
> > +  std::string Info() const;
> > +
> > + private:
> > +  unordered_map<int32, std::vector<int32> > binary_feats_;
> > +  unordered_map<int32, int32> phone_id_;
>
> OK, I misunderstood. So we will have a network which will do the
> avg-logprob-estimation and also will handle unseen-phones/backoff. I think
> this will be a good improvement of the design (of course if this
> subtraction of logprobs proves effective) in the sense that it is inherent
> and it is on GPU and handles backoff/unseen-phones automatically. But one
> problem is that we are training a whole separate network which might
> increase training time significantly. Also the context sizes that work well
> for duration modeling are quite bigger than biphone and triphone; Should we
> not test them for logprob-subtraction too to see if they are good? Or we
> could use a smaller sub-context for this separate neural net.
>
let's not worry about this until we know if having the context is even
important.

> I have not yet tested the subtraction with biphones. I will test this and
> 0-phone approximation. I probably have to test them on another dataset too
> (for eg. farsdat) to make sure they are useful.
>
ok cool.

> On another matter, the AMI experiment in which I used a separate root for
> each phone has not finished yet. That is because the input feature has
> become quite big (~1800) and so the rescoring (which is run on CPU) takes
> very long. I tried it 2 times and each time most of the 250 jobs were
> completed but a few (~10) passed maximum runtime of a job and were
> terminated. What should I do?
>
I don't believe we have a maximum runtime of a job o the CLSP cluster.  We
have had some grid issues in the last couple of days.  What was the error?

Dan




> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/414/files#r53530457>.
>